fix: Don't kill supervisord when processes exit, for #5906, fixes #5358

stasadev opened the pull request:
## The Issue

- #5358
- #5906
- #1142

## How This PR Solves The Issue

The logic added in #1142 was done before healthchecks were added to Docker containers.

The change proposed in #5906 makes the logic even more complex.

Alternatively, we can remove `kill_supervisor.py` completely.

## Manual Testing Instructions

Use this in `.ddev/config.yaml`:

```
web_extra_daemons:
  - name: baddaemon
    command: somemissingcommand
    directory: /var/www/html
  - name: gooddaemon
    command: 'while true; do date; sleep 60; done'
    directory: /var/www/html
```

```
ddev logs # see correct behavior for baddaemon
WARN exited: baddaemon (exit status 127; not expected)
```

```
ddev exec killall php-fpm
ddev logs # see it restarting again
INFO success: php-fpm entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
```

## Automated Testing Overview

<!-- Please describe the tests introduced by this PR, or explain why no tests are needed. -->

## Related Issue Link(s)

## Release/Deployment Notes

<!-- Does this affect anything else or have ramifications for other code? Does anything have to be done on deployment? -->

rfay commented:
This is working, and I think it's an improvement. However, we have a bit more work to do to make it really work right.

- [ ] Docs explaining how you see if a daemon is working (`ddev logs | grep INFO` for example, or `ddev logs | grep <daemonname>`)
- [ ] I'm pretty sure we have to do something with `exitcodes` (http://supervisord.org/configuration.html#program-x-section-settings) in the generated supervisord config; right now, a failure does *not* stop the restart behavior, the the output is not correct. 

I used this configuration in config.yaml:
```
web_extra_daemons:
  - name: baddaemon
    command: somemissingcommand
    directory: /var/www/html
  - name: gooddaemon
    command: 'while true; do date; sleep 60; done'
    directory: /var/www/html
```

It generates a config like this for baddaemon:
```
[program:baddaemon]
group=webextradaemons
command=bash -c "somemissingcommand || sleep 2"
directory=/var/www/html
autostart=false
autorestart=true
startretries=15
stdout_logfile=/var/tmp/logpipe
stdout_logfile_maxbytes=0
redirect_stderr=true
```

and then `ddev start` and `ddev logs`. I see:

```
2024-03-05 21:39:01,009 INFO success: mailpit entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-03-05 21:39:01,494 INFO spawned: 'baddaemon' with pid 2190
2024-03-05 21:39:01,495 INFO spawned: 'gooddaemon' with pid 2191
bash: line 1: somemissingcommand: command not found
Tue Mar  5 21:39:01 UTC 2024
2024-03-05 21:39:02,524 INFO success: baddaemon entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-03-05 21:39:02,525 INFO success: gooddaemon entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-03-05 21:39:03,523 INFO exited: baddaemon (exit status 0; expected)
2024-03-05 21:39:04,533 INFO spawned: 'baddaemon' with pid 2296
bash: line 1: somemissingcommand: command not found
2024-03-05 21:39:05,578 INFO success: baddaemon entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2024-03-05 21:39:06,576 INFO exited: baddaemon (exit status 0; expected)
2024-03-05 21:39:07,582 INFO spawned: 'baddaemon' with pid 2327
bash: line 1: somemissingcommand: command not found
2024-03-05 21:39:08,624 INFO success: baddaemon entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
```

We see that `baddaemon` is of course failing, and *not* staying up for more than 1 second. (Probably we need `startretries` > 1, perhaps 3?) And supervisord is saying that it *is* staying up, which isn't true.

rfay commented:
👍 Also, `ddev exec killall mailpit` kills mailpit and it gets restarted
However, apache, nginx, and php-fpm do not restart, and they'd be better that way, `autorestart=true`

stasadev commented:
I repushed the Docker tag:

```
docker pull ddev/ddev-webserver:20240305_stasadev_remove_kill_supervisor
```

> Docs explaining how you see if a daemon is working (`ddev logs | grep INFO` for example, or `ddev logs | grep <daemonname>`)

It's better to check the full output, and it's already described in the documentation, but it's not as noticeable as it should be, I made a small change for this.

> right now, a failure does not stop the restart behavior, the the output is not correct.
> We see that `baddaemon` is of course failing, and not staying up for more than 1 second. (Probably we need `startretries` > 1, perhaps 3?) And supervisord is saying that it is staying up, which isn't true.

We already had `startretries=15`, the problem was in `command=bash -c "somemissingcommand || sleep 2"`, `sleep` always returned 0.

I found the reason for `sleep 2`: [Use bash for web_extra_daemons command and add a sleep to wait for mutagen](https://github.com/ddev/ddev/pull/4057/commits/491c10203d9a9ac00a37d1aed69e35a07021e0cf).

I refactored it so that the command's return code was always returned, and added `startsecs=3` to wait 3 secs (sleep 2 + 1 extra sec) for success.

rfay commented on pkg/ddevapp/config.go:
```diff
@@ -948,10 +948,11 @@ func (app *DdevApp) RenderComposeYAML() (string, error) {
 [program:%s]
 group=webextradaemons
-command=bash -c "%s || sleep 2"
+command=bash -c "%s; exit_code=$?; if [ $exit_code -ne 0 ]; then sleep 2; fi; exit $exit_code"
```
Ah, this is also why it was reported as staying up more than 1s

rfay reviewed (approved):
Well, I really like it. I think it's more useful behavior all the way around, although it is definitely a change. 

Tested with the manual instructions and with `killall php-fpm` and `killall nginx` and everything came back and started working again.

rfay commented:
Rebased and pushed new image to pick up other image changes. Now with the test I see
```
Starting web_extra_daemons...
Unable to start web_extra_daemons using supervisorctl, stdout=webextradaemons:baddaemon: ERROR (spawn error)
webextradaemons:gooddaemon: started
, stderr=: composeCmd failed to run 'COMPOSE_PROJECT_NAME=ddev-d10 docker-compose -f /Users/rfay/workspace/d10/.ddev/.ddev-docker-compose-full.yaml exec -T web bash -c set -eu && ( supervisorctl start webextradaemons:*)', action='[]', err='exit status 7', stdout='webextradaemons:baddaemon: ERROR (spawn error)
webextradaemons:gooddaemon: started
', stderr=''
```

Which actually seems like a pretty good thing, but I don't think it's what I had before.

rfay commented:
It looks to me like this is all great, so pulling. 

To test, make sure to re-pull the latest 

```
docker pull ddev/ddev-webserver:20240305_stasadev_remove_kill_supervisor
```
